### ISSUE TITLE
[System.Text.Json] JsonSerializer ignores MaxDepth option

### ISSUE THREAD

**Issue by Groxan** [dotnet/runtime#882]
I caught an exception in production, and my service was stuck due to a bug in deserialization of objects with large depth using custom converters. I have created a sample repository which demonstrates the bug: https://github.com/Groxan/JsonSerializerDepthSample

In short, I have the following custom json converter:

````cs
class CustomConverter : JsonConverter<IContent>
{
    public override IContent Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        var sideReader = reader;

        sideReader.Read();
        sideReader.Read();
        var type = sideReader.GetString();

        return type switch
        {
            "array" => JsonSerializer.Deserialize<DeepArray>(ref reader, options),
            _ => throw new JsonException()
        };
    }

    public override void Write(Utf8JsonWriter writer, IContent value, JsonSerializerOptions options)
    {
        throw new NotImplementedException();
    }
}
````

And the code:

````cs
static void Main(string[] args)
{
    var json = File.ReadAllText("Sample.json");

    var options = new JsonSerializerOptions { MaxDepth = 256 };
    options.Converters.Add(new CustomConverter());

    // works fine
    var direct = JsonSerializer.Deserialize<DeepArray>(json, options);

    // fails with exception:
    // "The maximum configured depth of 64 has been exceeded. Cannot read next JSON array"
    var custom = JsonSerializer.Deserialize<IContent>(json, options);
}
````

So, when I call `JsonSerializer.Deserialize<T>` where `T` is the exact type, everything's fine. But when `T` is an interface and the custom converter is used, I catch an exception "The maximum configured depth of 64 has been exceeded", despite the fact that I set `MaxDepth = 256` in `JsonSerializerOptions`.

I suppose the `MaxDepth` parameter is simply ignored by the converter :(

**Comment by ahsonkhan** [comment]
> I suppose the `MaxDepth` parameter is simply ignored by the converter :(

This bug is not specific just to converters (nor just to `MaxDepth`) and will be hit whenever the caller passes in a `Utf8JsonReader` to the `JsonSerializer.Deserialize` method with non-default `JsonReaderOptions` set (which currently includes `MaxDepth`, `JsonCommentHandling`, and `AllowTrailingCommas`).

For example, this code snippet incorrectly fails with an exception:
```C#
var localReader = new Utf8JsonReader(Encoding.UTF8.GetBytes("[null,]"), 
                                       new JsonReaderOptions { AllowTrailingCommas = true });

string[] returnedValue = JsonSerializer.Deserialize<string[]>(ref localReader);

// System.Text.Json.JsonException : The JSON array contains a trailing comma at the end 
// which is not supported in this mode. Change the reader options. Path: $[1] | 
// LineNumber: 0 | BytePositionInLine: 6.
```

Here's the issue (note the new `Utf8JsonReader` is created without passing in the user defined options):
https://github.com/dotnet/runtime/blob/377b3c7c3e364b3183696744c6a5c6968bcdffd4/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Utf8JsonReader.cs#L302

Here's the fix (and requires us to add more tests where different options are exercised):
```C#
JsonReaderOptions originalReaderOptions = state.Options;

var newReader = new Utf8JsonReader(rentedSpan, originalReaderOptions);
```

We should consider fixing this in 3.1 as well.

**Comment by Groxan** [comment]
Thanks for your reply! I will be looking forward to...

**Comment by ahsonkhan** [comment]
Re-opening until fixed in 3.1

**Comment by layomia** [comment]
Fixed in 3.1 by https://github.com/dotnet/corefx/pull/42867.
